cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,101 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CloudPrivateIPConfig allows the user to assign private IP addresses
+// to the primary NIC associated with cloud VMs. This is done by
+// specifying a list Kubernetes nodes and IPs requested for those nodes.
+// This CRD is intended on being used by the network plugin which manages
+// the cluster network.
+// Note: this CRD is managed using a singleton cluster-wide CR, which is
+// called: `cluster`. This is enforced by the CRD's unique name in
+// combination with the fact that the CRD is cluster-scoped.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=cloudprivateipconfig,scope=Cluster
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// spec is the definition of the desired private IP request(s).
+	// +kubebuilder:validation:Required
+	// +required
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec,casttype=CloudPrivateIPConfigSpec"`
+	// status is the observed status of the desired private IP request(s). Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status,casttype=CloudPrivateIPConfigStatus"`
+}
+
+// CloudPrivateIPConfigSpec specifies the requested list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpec struct {
+	// +kubebuilder:validation:Required
+	// +required
+	Items []CloudPrivateIPConfigSpecItem `json:"items" protobuf:"bytes,1,opt,name=items"`
```

Comment:
create a little bit of doc here. does order matter?  We use single-instances for cluster-wide configuration in config.openshift.io because it is largely "set this single value" style API.  What is driving the choice of cluster for this API?

Reply:
missing restrictions on duplicate names?  What if I set the same node value multiple times?

Reply:
> does order matter?

No

>  What is driving the choice of a single instance called cluster for this API?

TL;DR: the answer is "atomicity of update is required"

This API is helper for implementing the OpenShift specific networking feature: egress IP, on regular clouds (AWS/GCP/Azure). Egress IP has the particularity of not having "static" IP address assignments, which means: an IP address 192.168.126.12 can be assigned to node A right now, and be moved to node B later. When such a change happens we can't have node A and node B claiming 192.168.126.12 at any moment in time. Hence, as to avoid that we'll use a single CR instance making sure the IP is moved from node A to node B atomically. This comment re-phrases what I just said (but maybe it makes more sense?): https://github.com/openshift/enhancements/pull/571#discussion_r587709378

> What if I set the same node value multiple times?

Not a problem. node A could host several different IPs. However, the tuple: node<->IP needs to be unique, I can't seem to enforce that though since if I do:

```
// +kubebuilder:validation:UniqueItems=true
// +kubebuilder:validation:Required
// +required
Items []CloudPrivateIPConfigSpecItem `json:"items" protobuf:"bytes,1,opt,name=items"`
```

I get a message 

```
The CustomResourceDefinition "cloudprivateipconfig.cloud.network.openshift.io" is invalid: spec.validation.openAPIV3Schema.properties[status].properties[items].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic
```

So I was thinking to enforce it in the controller / admission controller.

Reply:
Not sure if it was made explicit anywhere here, but this is an object that is only written to by operators, not by users (even administrators). The `spec` side represents the current state that the network plugin wants, and the `status` side represents the current state that the "cloud egress IP operator" has implemented. No users will have permission to modify it, and if a cluster-admin decides to edit it for some reason, they can't do anything _useful_ with it (and their changes will be overwritten the next time the operators update the object).

Reply:
if it needs atomicity, then why not just use the IP as the resource name?

Reply:
> if it needs atomicity, then why not just use the IP as the resource name?

We could do that too. It's not been mentioned here because no one mentioned it on the enhancement proposal, or the meetings we had prior to that. But the proposal mentioned in the description of this PR would work too.

Reply:
Sorry about that my mistake. I just fell off the enhancement discussion.